### PR TITLE
BLOCKCHAIN-387 FE DEX 

### PR DESCRIPTION
**Pull request by MichalLustykNeti**

### CODE REVIEW

File: src/api/nodeRpcCall.js [hunk 1]
```diff
@@ -1924,6 +1997,115 @@ const fetchPendingIdentities = async () => {
...
+  lpTokenBurn,
+  amount1MinReceive,
+  amount2MinReceive,
+  withdrawTo,
+  walletAddress,
+) => {
+  const api = await getApi();
+  const extrinsic = api.tx.assetConversion.removeLiquidity(
+    asset1,
+
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:2094]:
```suggestion
```

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Review by kacperzuk-neti — APPROVED** [review]

**Comment by MichalLustykNeti** [comment]
image
If pool exist but there are no liquidity in it

**Comment by MichalLustykNeti** [comment]
for 
asset1=1000000000000
asset2=1000000000000
image
reserves shouldn't be zero

**Comment by MichalLustykNeti** [comment]
if there are not enough funds in reserves we should let user know how many token max he may swap. For now i have to guess how many funds can i swap
image

File: src/components/Wallet/Exchange/ExchangeShowMore/index.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        <div>
+          {`In the pool
+              ${formatProperlyValue(asset1, reserved.asset1, asset1ToShow, asset1Decimals || 0)}  
+            / ${formatProperlyValue(asset2, reserved.asset2, asset2ToShow, asset2Decimals || 0)}` }
+        </div>
+        )}
+      </div>
+      <div className={styles.liquidity}>
+        {liquidity
+          ? `Your liquidity: ${((liquidity / lpTokensBalance) * 100)}% (${liquidity} Lp Tokens)`
```
Comment (kacperzuk-neti) [inline src/components/Wallet/Exchange/ExchangeShowMore/index.js:38]: % isn't really important. In a healthy system, all users will have some very small share of the LP.
Number of LP tokens similarly isn't really useful to a user.
What is useful is how many asset1 and asset2 "belongs" to the user.

See https://support.uniswap.org/hc/article_attachments/21158962149645 for an example of how it looks in uniswap v2, on which our dex is loosely based.

File: src/api/nodeRpcCall.js [hunk 3]
```diff
@@ -1925,6 +2029,191 @@ const fetchPendingIdentities = async () => {
...
+const getDexPools = async (walletAddress) => {
+  try {
+    const api = await getApi();
+    const pools = await api.query.assetConversion.pools.entries();
+    const poolsData = await Promise.all(pools.map(async ([poolKey, maybePoolData]) => {
+      const [asset1, asset2] = poolKey.args[0];
+      const { lpToken } = maybePoolData.unwrapOrDefault();
+      const asset1checkIsNative = asset1.value.toString();
+      const asset2checkIsNative = asset2.value.toString();
+      const asset2Tronsform = asset2checkIsNative || asset2.toString();
```
Comment (kacperzuk-neti) [inline src/api/nodeRpcCall.js:2095]:
```suggestion
      const asset2Transform = asset2checkIsNative || asset2.toString();
```
This typo shows up in other places too, please fix all.

File: src/components/Wallet/Exchange/ExchangeShowMore/index.js (hunk 2 above)
Comment (kacperzuk-neti) [inline src/components/Wallet/Exchange/ExchangeShowMore/index.js:38]: I have trouble tracing what `lpTokensBalance` is. However, we should divide by any balance here - we should divide by total supply of these tokens.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Comment by DorianSternVukotic** [comment]
Conflicts

File: src/redux/reducers/dex.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    }),
+
+    [dexActions.getPools.success]: (state, action) => ({
+      ...state,
+      pools: action.payload,
+    }),
+
+    [dexActions.getDexReserves.success]: (state, action) => ({
+      ...state,
+      reserves: { [action.payload.asset1Number + action.payload.asset2Number]: action.payload, ...state.reserves },
```
Comment (kacperzuk-neti) [inline src/redux/reducers/dex.js:52]: Won't this result in conflicts? What if we have pool for asset numbers `1` and `4` and a pool for assets `2` and `3`? They both will sum to key `5`.

File: src/utils/dexFormater.js [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+  const amount2Min = calculateAmountMin(
+    amount2,
+    minAmountPercent,
+  );
+  return {
+    amount1, amount2, amount1Min, amount2Min,
+  };
+};
+
+export const converTransferData = async (
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:143]:
```suggestion
export const convertTransferData = async (
```

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/redux/reducers/dex.js (hunk 4 above)
Comment (SzymonZiebaNeti) [reply src/redux/reducers/dex.js:52]: asset1Number should be string, I want to make strings 14 and 23, but i will check if there is always string

Comment (kacperzuk-neti) [reply src/redux/reducers/dex.js:52]: That doesnt change it. If asset1 is `1` and asset2 is `21` it will conflict with assets `12` and `1`

File: src/utils/dexFormater.js [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  const stringValue = mod.toString();
+
+  const integerPart = stringValue.slice(0, -1e18.toString().length) || '0';
+  const decimalPart = stringValue.slice(-1e18.toString().length);
+
+  const dividedValueString = `${integerPart}.${decimalPart}`;
+  const substract = new BN(Number(div) + Number(dividedValueString));
+  const result = subResult
+    ? amount.sub(substract.isZero() ? new BN(2) : substract)
+    : amount.add(substract.isZero() ? new BN(2) : substract);
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:100]: What this code does?

File: src/utils/dexFormater.js [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+  return result;
+};
+
+export const calculateAmountDesiredFormatted = (desiredAmount, decimals) => {
+  const valueString = desiredAmount.toString();
+  const decimalsFromValue = valueString.split('.')[1]?.length || '0';
+  return new BN(valueString.replace('.', ''))
+    .mul(decimals)
+    .div((new BN(10)
+      .pow(new BN(decimalsFromValue))));
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:110]: Don't reinvent parsing balances, use `parseAssets` from walletHelpers.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]
Parsing/formatting balances (decimal handling) logic needs to be refactored.
1. Most of it should be done using `parseAssets` and `formatAssets` from `walletHelpers`
2. For others there's still way to many places that do random `replace('.', '')`, `split('.')` or `pow(decimals)`. Rethink this, find common abstractions and refactor. Right now I guarantee that there's a dozen different subtle bugs in there.

File: src/components/Modals/TradeTokens.js [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  }, [asset1, asset2, reserves]);
+
+  const {
+    handleSubmit,
+    register,
+    formState: { errors, isValid },
+  } = useForm({
+    mode: 'onChange',
+    defaultValues: {
+      cos: null,
```
Comment (kacperzuk-neti) [inline src/components/Modals/TradeTokens.js:67]: Some old placeholder?

File: src/components/Modals/TradeTokens.js [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+      );
+      const swapData = {
+        path: { asset1, asset2 },
+        amount,
+        amountMin,
+        sendTo: walletAddress,
+      };
+      dispatch(
+        isBuy
+          ? dexActions.swapTokensForExactTokens.call(swapData) : dexActions.swapExactTokensForTokens.call(swapData),
```
Comment (kacperzuk-neti) [inline src/components/Modals/TradeTokens.js:93]: I think there might be some misunderstanding here.
Both this calls do the same kind of trade - sell asset1 in exchange for asset2. It's just that with `swapTokensForExactTokens` is then you're saying "I sell asset1, I get 100 asset2" and `swapExactTokensForTokens` is "I sell 100 asset1, I get asset2". To actually sell asset2 and buy asset1 you need to swap them in the `path`.

Right now UI shows only "Buy LLD for Asset" and "Sell LLD For Asset". This isn't enough. In general I think that using custom buttons for this plainly sux UX-wise and even if us developers get confused our users have no chance and we should just copy Uniswap, buuut if we want to keep buttons, then we need 4 of them (there's a Polish CEX (zonda) that kinda does it like that, though with tabs instead of buttons):
* `Buy LLD for Asset` - swapTokensForExactTokens, where Asset is asset1 and LLD is asset2
* `Buy Asset for LLD` - swapTokensForExactTokens, where LLD is asset1 and Asset is asset2
* `Sell LLD for Asset` - swapExactTokensForTokens, where LLD is asset1 and Asset is asset2
* `Sell Asset for LLD` - swapExactTokensForTokens, where Asset is asset1 and LLD is asset2

In other words, `asset1` is always what user will lose, `asset2` is what user will gain.  Whether we use swapTokensForExactTokens or swapExactTokensForTokens depends on which amount user wants to specify, but doesn't affect the direction of the trade.

@DorianSternVukotic this is why you can't fix the ratio.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/components/Modals/TradeTokens.js (hunk 9 above)
Comment (kacperzuk) [reply src/components/Modals/TradeTokens.js:93]: After signal chat, we keep 2 buttons "Buy LLD for Asset" and "Sell LLD for Asset" and they decide the direction of trade (which asset is asset1 and which is asset2). For the forms, we do two inputs: one for how much we pay (asset1) and one for how much we receive (asset2). Then, depending on which input the user puts amount in, we decide on which side is exact (pay or receive). So:
1. `Buy LLD for Asset` - Asset is asset1, LLD is asset2
2. `Sell LLD for Asset` - LLD is asset1, Asset is asset2
3. If user puts amount in the first input (the "I lose/pay with" input), we do `swapExactTokensForTokens`
4. If user puts amount in the second input (the "I buy/receive"), we do `swapTokensForExactTokens`

File: src/components/Modals/AddLiquidityModal.js [hunk 10]
```diff
@@ -87,55 +78,41 @@ function AddLiquidityModal({
...
+    const sanitizedValue = sanitizeValue(value.toString());
+    const assetFormat = isAsset1 ? asset2 : asset1;
+    const amount = parseProperlyValue(
+      isAsset1 ? asset1 : asset2,
+      sanitizedValue,
+      isAsset1 ? assetData1?.decimals : assetData2?.decimals,
+    );
+    const tradeData = await (isAsset1
+      ? getSwapPriceExactTokensForTokens
+      : getSwapPriceTokensForExactTokens)(enum1, enum2, amount, false);
```
Comment (kacperzuk-neti) [inline src/components/Modals/AddLiquidityModal.js:107]:
```suggestion
    const getSwapPrice = isAsset1 ? getSwapPriceExactTokensForTokens : getSwapPriceTokensForExactTokens;
    const tradeData = await getSwapPrice(enum1, enum2, amount, false);
```
Will be much more readable.

File: src/utils/dexFormater.js [hunk 11]
```diff
@@ -55,21 +60,33 @@ export const makeAssetToShow = (asset, symbol) => {
...
   if (asset === 'Native') {
-    formattedValue = formatDollars(value);
+    formattedValue = formatDollars(value, true);
   } else if (asset === '1') {
-    formattedValue = formatMerits(value);
+    formattedValue = formatMerits(value, true);
   } else {
-    formattedValue = formatter(value, decimals);
+    formattedValue = formatAssets(value, decimals || 0, true);
   }
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js]: If we have decimals, then we can just use `formatAssets` for everything.

File: src/utils/dexFormater.js [hunk 12]
```diff
@@ -55,21 +60,33 @@ export const makeAssetToShow = (asset, symbol) => {
   return assetToShow;
 };
 
-export const formatProperlyValue = (asset, value, symbol, decimals) => {
+export const formatProperlyValue = (asset, value, decimals, symbol) => {
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:63]: Terrible name. Maybe `formatAssetsWithSymbol` and move to `walletHelpers`? Or even better. Change the formatAssets in walletHelpers to optionally accept symbol and add it if provided.

File: src/utils/dexFormater.js [hunk 13]
```diff
@@ -55,21 +60,33 @@ export const makeAssetToShow = (asset, symbol) => {
...
 
+export const parseProperlyValue = (asset, value, decimals) => {
+  let parsedValue;
+  if (asset === 'Native') {
+    parsedValue = parseDollars(value);
+  } else if (asset === '1') {
+    parsedValue = parseMerits(value);
+  } else {
+    parsedValue = parseAssets(value, decimals);
+  }
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:85]: If we have decimals, then we can just call `parseAssets(value, decimals)` even for LLD and LLM, which makes `parseProperlyValue` totally redundant.

File: src/utils/dexFormater.js [hunk 14]
```diff
@@ -85,47 +102,30 @@ const calculateAmountMin = (
...
-  const decimalsFromValue = valueString.split('.')[1]?.length || '0';
-  return new BN(valueString.replace('.', ''))
-    .mul(decimals)
-    .div((new BN(10)
-      .pow(new BN(decimalsFromValue))));
+  const minAmountValue = (Number(minAmountPercent || defaultMinPercent) / 100).toString();
+  const lengthDecimalPercent = minAmountValue.split('.')[1].length;
+  const minAmountPercentBN = parseAssets(minAmountValue, lengthDecimalPercent);
+  const decimalsBN = new BN(10).pow(new BN(lengthDecimalPercent));
+  const percentBN = amount.mul(minAmountPercentBN).divn(decimalsBN);
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:109]: Way too overengineered. Just assume a big enough denominator instead of trying to guess it based on string formatting of the fraction. So something like this:
```
const denominator = 10000;
const slippagePercentBN = new BN(minAmountPercent * denominator / 100);
const maxSlippage = amount.mul(slippagePercentBN).div(denominator);
```

File: src/utils/dexFormater.js [hunk 15]
```diff
@@ -85,47 +102,30 @@ const calculateAmountMin = (
...
+
+  if (subResult) {
+    const sub = amount.sub(percentBN.isZero() ? new BN(3) : percentBN);
+    return sub.lte(new BN(0)) ? 1 : sub;
+  }
+  return amount.add(percentBN.isZero() ? new BN(3) : percentBN);
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:115]: I assume that the `add` case is actually for calculating `amountMax`, not `amountMin`. This doesn't belong in a single function. Call this function (that calculates the slippage) `calculateSlippage` and do a separate `calculateAmountMin` that subs it and `calculateAmountMax` that adds it.

Also it's not a formatter in any way, it's business logic, so doesn't belong in `dexFormater.js`

File: src/utils/dexFormater.js [hunk 16]
```diff
@@ -1,7 +1,12 @@
 import { formatBalance, BN } from '@polkadot/util';
```
Comment (kacperzuk-neti) [inline src/utils/dexFormater.js:1]: Typo in filename - should be `dexFormatter.js`

File: src/components/Modals/AddLiquidityModal.js (hunk 10 above)
Comment (kacperzuk-neti) [reply src/components/Modals/AddLiquidityModal.js:107]: The same pattern is repeated in multiple places btw.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/api/nodeRpcCall.js [hunk 17]
```diff
@@ -5,9 +5,8 @@ import { USER_ROLES, userRolesHelper } from '../utils/userRolesHelper';
 import { handleMyDispatchErrors } from '../utils/therapist';
 import { blockchainDataToFormObject } from '../utils/registryFormBuilder';
 import * as centralizedBackend from './backend';
-import {
-  convertAssetData,
-} from '../utils/dexFormater';
+// eslint-disable-next-line import/no-cycle
```
[the remark this answers is not in the document]
Comment (SzymonZiebaNeti) [reply src/api/nodeRpcCall.js:8]: It's because in dexFormatter we have import from nodeRpcCall, and in nodeRpcCall i have import from dexFormatter? should i just move this function convertAssetData to nodeRpcCall?

[the remark this answers is not in the document]
Comment (kacperzuk-neti) [reply src/api/nodeRpcCall.js:8]: In perfect world I'd keep `nodeRpcCall` as a very thin wrapper over actual APIs and not put business logic in here at all. `convertAssetData` is something that fits more in sagas.

But I don't care much, there's a lot of business logic in here either way. I just missed the import of `nodeRpcCall` in `dexFormatter` and was genuinely confused as to why we need to silence eslint here. If it's a real cycle and it causes no issues I'm fine leaving it as-is.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

**Review by kacperzuk-neti — APPROVED** [review]

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/components/Modals/AddLiquidityModal.js [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+    const formatedValue = formatAssets(tradeData, decimalsOut);
+    const sanitizedValue = sanitizeValue(formatedValue.toString());
+    setValue('amount1Desired', isAsset1 ? value : sanitizedValue);
+    setValue('amount2Desired', isAsset1 ? sanitizedValue : value);
+    await trigger();
+  };
+
+  useEffect(() => {
+    dispatch(walletActions.getAssetsBalance.call([asset1, asset2]));
+  }, [dispatch, asset1, asset2]);
```
[the remark this answers is not in the document]
Comment (kacperzuk-neti) [reply src/components/Modals/AddLiquidityModal.js:121]: Not needed, this is modal.

File: src/components/Modals/TradeTokens.js [hunk 19]
```diff
@@ -0,0 +1,354 @@
...
+    return true;
+  };
+
+  useEffect(() => {
+    dispatch(dexActions.getDexReserves.call({ asset1, asset2 }));
+  }, [dispatch, asset1, asset2]);
+
+  useEffect(() => {
+    dispatch(walletActions.getAssetsBalance.call([asset1, asset2]));
+  }, [dispatch, asset1, asset2]);
```
[the remark this answers is not in the document]
Comment (kacperzuk-neti) [reply src/components/Modals/TradeTokens.js:227]: Not needed, this is modal.

File: src/components/Modals/RemoveLiquidity/index.js [hunk 20]
```diff
@@ -46,14 +47,14 @@ function RemoveLiquidityModal({
...
+    const tokensToBurn = new BN(lpTokensBalance).mul(new BN(numberValue)).div(BN_HUNDRED);
     setTokensToBurnState(tokensToBurn);
-    const fee = 1; // 0.1%
+    const fee = new BN(withdrawlFee);
     const calculatedAmount1 = calculatePooled(tokensToBurn, liquidity, reserved.asset1);
     const calculatedAmount2 = calculatePooled(tokensToBurn, liquidity, reserved.asset2);
-    const asset1Data = calculatedAmount1.mul(new BN(BN_THOUSAND - fee)).div(new BN(BN_THOUSAND));
-    const asset2Data = calculatedAmount2.mul(new BN(BN_THOUSAND - fee)).div(new BN(BN_THOUSAND));
+    const asset1Data = !fee.isZero() ? calculatedAmount1.sub(calculatedAmount1.div(fee)) : calculatedAmount1;
+    const asset2Data = !fee.isZero() ? calculatedAmount2.sub(calculatedAmount2.div(fee)) : calculatedAmount2;
```
Comment (kacperzuk-neti) [inline src/components/Modals/RemoveLiquidity/index.js:57]: Why `amount - amount/fee`? Shouldn't it be `amount * (1000-fee)/1000` or `amount - amount * fee/1000`?

File: src/redux/actions/dex.js [hunk 21]
```diff
@@ -45,4 +46,9 @@ export const {
     SUCCESS: undefined,
     FAILURE: undefined,
   },
+  GET_WITHDRAWL_FEE: {
```
Comment (kacperzuk-neti) [inline src/redux/actions/dex.js:49]:
```suggestion
  GET_WITHDRAWAL_FEE: {
```
Typo.

**Review by kacperzuk-neti — CHANGES_REQUESTED** [review]

File: src/components/Modals/RemoveLiquidity/index.js [hunk 22]
```diff
@@ -0,0 +1,196 @@
...
+    setTokensToBurnState(tokensToBurn);
+    const fee = new BN(withdrawalFee);
+    const calculatedAmount1 = calculatePooled(tokensToBurn, liquidity, reserved.asset1);
+    const calculatedAmount2 = calculatePooled(tokensToBurn, liquidity, reserved.asset2);
+    const asset1Data = !fee.isZero()
+      ? calculatedAmount1.sub(calculatedAmount1.mul(fee).div(BN_MILLION))
+      : calculatedAmount1;
+    const asset2Data = !fee.isZero()
+      ? calculatedAmount2.sub(calculatedAmount2.mul(fee).div(BN_MILLION))
+      : calculatedAmount2;
```
Comment (kacperzuk-neti) [inline src/components/Modals/RemoveLiquidity/index.js:61]:
```suggestion
    const asset1Data = calculatedAmount1.sub(calculatedAmount1.mul(fee).div(BN_MILLION));
    const asset2Data = calculatedAmount2.sub(calculatedAmount2.mul(fee).div(BN_MILLION));
```
Multiplication will handle `fee = 0` just fine

**Review by kacperzuk-neti — APPROVED** [review]

**Review by DorianSternVukotic — CHANGES_REQUESTED** [review]
